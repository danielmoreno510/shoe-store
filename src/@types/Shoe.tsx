type ShoeProps = {
  shoe: Shoe;
  cart?: Shoe[];
  addToCart?(shoe: Shoe): any;
  showProduct?(shoe: Shoe): any;
  twoSection?: boolean;
  oneSection?: boolean;
  newSize?: number;
};
