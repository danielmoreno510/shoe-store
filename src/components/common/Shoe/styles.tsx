import {StyleSheet} from 'react-native';

import {primaryColor} from '../../../constants';

export default StyleSheet.create({
  shoe: {
    borderBottomColor: primaryColor,
    borderBottomWidth: 1,
    flexDirection: 'row',
    alignContent: 'center',
    padding: 20,
    paddingHorizontal: 0,
    justifyContent: 'space-between',
  },
  shoeInfoContainer: {
    flex: 1,
  },
  carouselContainer: {
    alignItems: 'center',
    marginBottom: -20,
  },
  reference: {
    fontSize: 22,
    fontWeight: '500',
  },
  brand: {
    fontSize: 18,
  },
  price: {
    fontSize: 28,
    color: primaryColor,
    fontWeight: 'bold',
    marginTop: 12,
  },
  infoContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  oneInfoContainer: {
    justifyContent: 'space-between',
    flexDirection: 'row',
    flex: 1,
    alignItems: 'center',
  },
  sectionContainer: {
    flex: 1,
    alignItems: 'center',
  },
  sizeContainer: {
    flexDirection: 'row',
    alignContent: 'center',
    marginTop: 12,
  },
  quantityContainer: {
    flexDirection: 'row',
  },
  size: {
    fontSize: 20,
  },
  quantity: {
    fontSize: 20,
  },
  sizeSelectContainer: {
    borderWidth: 1,
    borderColor: primaryColor,
    width: 40,
    borderRadius: 4,
  },
  buttonsContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingTop: 12,
  },
  button: {
    margin: 12,
    width: 54,
  },
  icon: {
    color: 'white',
  },
  image: {
    width: 180,
    height: 180,
  },
});

export const pickerSelectStyles = StyleSheet.create({
  inputIOS: {
    fontSize: 20,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  inputAndroid: {
    fontSize: 20,
    fontWeight: 'bold',
    textAlign: 'center',
  },
});
