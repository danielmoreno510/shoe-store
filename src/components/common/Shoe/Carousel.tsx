import React from 'react';
import {View, Image} from 'react-native';
import Carousel, {Pagination} from 'react-native-snap-carousel';

import styles from './styles';
import shoes from '../../../assets/images/shoes';
import {primaryColor} from '../../../constants';

export const ShoeImage: React.FC<ShoeImageProps> = ({item, size}) => (
  <Image style={[{width: size, height: size}]} source={shoes[item]} />
);

const CustomCarousel: React.FC<CarouselProps> = ({shoe, width = 180}) => {
  const [currentImage, setCurrentImage] = React.useState<number>(0);

  return (
    <View style={styles.carouselContainer}>
      <Carousel
        data={shoe.images}
        renderItem={({item}) => <ShoeImage item={item} size={width} />}
        sliderWidth={width}
        itemWidth={width}
        inactiveSlideOpacity={0.85}
        autoplayDelay={500}
        autoplayInterval={3000}
        onSnapToItem={index => setCurrentImage(index)}
        autoplay={true}
        loop
      />
      <Pagination
        dotsLength={shoe.images.length}
        activeDotIndex={currentImage}
        dotColor={primaryColor}
        inactiveDotColor="black"
        inactiveDotOpacity={0.4}
        inactiveDotScale={0.6}
      />
    </View>
  );
};

export default CustomCarousel;
