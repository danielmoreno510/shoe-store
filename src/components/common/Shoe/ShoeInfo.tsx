import React from 'react';
import {View, Text} from 'react-native';
import {Icon} from 'native-base';
import RNPickerSelect from 'react-native-picker-select';
import {filter, find} from 'lodash';

import Button from '../Button';
import styles, {pickerSelectStyles} from './styles';

const ShoeInfo: React.FC<ShoeProps> = ({
  shoe,
  cart,
  addToCart,
  showProduct,
  twoSection,
  oneSection,
  newSize,
}) => {
  const [size, setSize] = React.useState<number>(4.5);
  const getQuantity = () => {
    const quantityShoes = filter(cart, ['id', shoe.id]);
    const shoesAvailable = find(shoe.sizes, ['value', size]);
    return shoesAvailable.quantity - quantityShoes.length;
  };

  return (
    <View style={[twoSection && styles.infoContainer, oneSection && styles.oneInfoContainer]}>
      <View style={styles.sectionContainer}>
        <Text style={styles.reference}>{shoe.reference}</Text>
        <Text style={styles.brand}>{shoe.brand}</Text>
        <View style={styles.sizeContainer}>
          <Text style={styles.size}>Size: </Text>
          {!newSize ? (
            <View style={styles.quantityContainer}>
              <View style={styles.sizeSelectContainer}>
                <RNPickerSelect
                  onValueChange={value => setSize(value)}
                  items={shoe.sizes}
                  value={size}
                  style={pickerSelectStyles}
                />
              </View>
              <Text style={styles.quantity}>{` ${getQuantity()}U`}</Text>
            </View>
          ) : (
            <Text style={styles.size}>{newSize}</Text>
          )}
        </View>
      </View>
      <View style={[styles.sectionContainer]}>
        <Text style={styles.price}>{`$${shoe.price}`}</Text>
        <View style={styles.buttonsContainer}>
          {addToCart && getQuantity() > 0 && (
            <Button testID="addToCart" onPress={() => addToCart({...shoe, size})} styleButton={styles.button}>
              <Icon type="MaterialIcons" name="shopping-cart" style={styles.icon} />
            </Button>
          )}
          {showProduct && (
            <Button testID="showProduct" onPress={() => showProduct(shoe)} styleButton={styles.button}>
              <Icon type="MaterialIcons" name="visibility" style={styles.icon} />
            </Button>
          )}
        </View>
        {!oneSection && <Text>{`Date: ${shoe.date}`}</Text>}
      </View>
    </View>
  );
};

export default ShoeInfo;
