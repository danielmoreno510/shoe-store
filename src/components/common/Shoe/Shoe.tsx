import React from 'react';
import {View} from 'react-native';

import Carousel from './Carousel';
import ShoeInfo from './ShoeInfo';
import styles from './styles';

const Shoe: React.FC<ShoeProps> = ({shoe, cart, addToCart, showProduct}) => (
  <View style={styles.shoe}>
    <Carousel shoe={shoe} />
    <View style={styles.shoeInfoContainer}>
      <ShoeInfo shoe={shoe} cart={cart} addToCart={addToCart} showProduct={showProduct} />
    </View>
  </View>
);

export default Shoe;
