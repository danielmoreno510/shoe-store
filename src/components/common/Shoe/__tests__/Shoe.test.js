import React from 'react';
import {shallow} from 'enzyme';

import Shoe from '../Shoe';

describe('Shoe component', () => {
  describe('GIVEN a set of valid props', () => {
    const testProps = {
      shoe: {},
      cart: [],
      addToCart: jest.fn(),
      showProduct: jest.fn(),
    };

    describe('WHEN the component is rendered', () => {
      const shallowWrapper = shallow(<Shoe {...testProps} />);

      it('THEN should display a regular Shoe', () => {
        expect(shallowWrapper).toMatchSnapshot();
      });
    });
  });
});
