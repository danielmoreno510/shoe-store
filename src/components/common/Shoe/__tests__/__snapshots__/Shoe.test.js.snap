// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shoe component GIVEN a set of valid props WHEN the component is rendered THEN should display a regular Shoe 1`] = `
<View
  style={
    Object {
      "alignContent": "center",
      "borderBottomColor": "#f1b30a",
      "borderBottomWidth": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "padding": 20,
      "paddingHorizontal": 0,
    }
  }
>
  <CustomCarousel
    shoe={Object {}}
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ShoeInfo
      addToCart={[MockFunction]}
      cart={Array []}
      shoe={Object {}}
      showProduct={[MockFunction]}
    />
  </View>
</View>
`;
